### PR TITLE
[WIP] refactor: resource table

### PR DESCRIPTION
**Pull request by bartlomieju**
Towards #3134

### CODE REVIEW

File: cli/resources.rs [hunk 1]
```diff
@@ -489,97 +633,61 @@ impl Future for ChildStatus {
...
-      maybe_std_file_copy.map_err(ErrBox::from)
-    }
-    _ => Err(bad_resource()),
-  }
+  let mut repr = table.remove(&rid).ok_or_else(bad_resource)?;
+  let fs_file = repr.downcast::<NewFsFile>().ok_or_else(bad_resource)?;
+  // Trait Clone not implemented on tokio::fs::File,
+  // so convert to std File first.
+  // TODO:
+  return Err(bad_resource());
```
Comment (bartlomieju) [inline cli/resources.rs]: I'm not sure if downcasting will work in this case, we need to take ownership of the file.

File: cli/resources.rs [hunk 2]
```diff
@@ -247,21 +190,47 @@ impl DenoAsyncRead for Resource {
   fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, ErrBox> {
     let mut table = RESOURCE_TABLE.lock().unwrap();
     let repr = table.get_mut(&self.rid).ok_or_else(bad_resource)?;
```
Comment (bartlomieju) [inline cli/resources.rs]: I'm still iterating on the API but I found this bit tricky. I'd like to be able to downcast `dyn DenoResource` to `DenoAsyncRead`, but `DenoAsyncRead` is a trait object it's not possible.
```
// what I'd like to do:
  resources::with_mut_resource(&self.rid, move |resource| {
      let resource = resource
        .downcast_mut::<dyn DenoAsyncRead>()
        .ok_or_else(bad_resource)?;
      f.0.poll_read(buf).map_err(ErrBox::from)
    })
```

@piscisaureus any advice?

**Comment by ry** [comment]
@bartlomieju Can we do the core version of this before cli? It's easier to think about API in core.

**Comment by ry** [comment]
@bartlomieju please open a new PR when you do cli. Closing this one just to keep the PR list small. Thanks.

**Pull request closed without merging**